adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,175 @@
+    private final Map<SubscriptionKey, TriTuple<CommandConsumer, LoraProvider, Context>> commandSubscriptions;
+
+    /**
+     * Creates a new Lora command subscriptions class.
+     *
+     * @param vertx The Vert.x instance to use.
+     */
+    public LoraCommandSubscriptions(final Vertx vertx) {
+        this(vertx, null);
+    }
+
+    /**
+     * Creates a new Lora command subscriptions class.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param tracer The tracer instance.
+     */
+    public LoraCommandSubscriptions(final Vertx vertx, final Tracer tracer) {
+        this.commandSubscriptions = new ConcurrentHashMap<>();
+        this.vertx = vertx;
+        this.tracer = tracer == null ? NoopTracerFactory.create() : tracer;
+    }
+
+    /**
+     * Subscribes for tenant diasbled/deleted events, in order to close the consumer objects associated with them later.
+     */
+    public void init() {
+        NotificationEventBusSupport.registerConsumer(vertx, TenantChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange())
+                                    && !notification.isTenantEnabled())) {
+                        closeConsumersForTenant(notification.getTenantId());
+                    }
+                });
+
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())) {
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final TriTuple<CommandConsumer, LoraProvider, Context> subscription = commandSubscriptions.remove(key);
+        if (subscription == null) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscription for deleted gateway",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final Context ctx = subscription.three();
+        ctx.runOnContext(x -> {
+            final CommandConsumer commandConsumer = subscription.one();
+            consumerCloseFutures.add(commandConsumer.close(null));
+        });
+        CompositeFuture.join(consumerCloseFutures).onComplete(x -> span.finish());
```

Comment:
I think it is probably better to finish the span here on the same thread/vert.x-context as the one on which the span was created (using `runOnContext`).

Reply:
But the span isn't created on the verticle context, only the consumer is closed there, right?

Reply:
The span is created on the context on which the event bus message was received (coming from `vertx.getOrCreateContext()` invoked when creating the event bus consumer in the `init` method). So, yes, it's different from the contexts on which the adapter verticles operate.

Reply:
Or, since there is only _one_ `commandConsumer.close()` invocation with corresponding LoraAdapter vertx context involved here, the span creation can be moved into the `ctx.runOnContext(` block as well.